src/mbedtls/pk.pyx

```diff
@@ -1036,3 +1037,76 @@ cdef class ECDHClient(ECDHBase):
         cdef const unsigned char* end = &buffer[-1] + 1
         check_error(_pk.mbedtls_ecdh_read_params(
             &self._ctx, &first, end))
+
+
+cdef class ECDHNaive(ECDHBase):
+
+    """Naive ECDH key exchange.
+
+    Args:
+        (Curve, optional): b'curve25519' or b'curve448'.
+
+    """
+    def __init__(self, curve=None):
+        super().__init__()
+        self.curve = curve or b'curve25519'
+        if self.curve == b'curve25519':
+            check_error(mbedtls_ecp_group_load(
+                &self._ctx.grp, _pk.MBEDTLS_ECP_DP_CURVE25519))
+        elif self.curve == b'curve448':
+            check_error(mbedtls_ecp_group_load(
+                &self._ctx.grp, _pk.MBEDTLS_ECP_DP_CURVE448))
+        else:
+            raise ValueError(
+                'ECDHNaive only supports curve25519 and curve448')
+
+    def generate(self):
+        """Generate a public key.
+
+        Return:
+            bytes: public key.
+
+        """
+        cdef unsigned char* output = <unsigned char*>malloc(
+            _mpi.MBEDTLS_MPI_MAX_SIZE * sizeof(unsigned char))
+        cdef size_t olen = 0
+        if not output:
+            raise MemoryError()
+        try:
+            check_error(_pk.mbedtls_ecdh_gen_public(
+                &self._ctx.grp, &self._ctx.d, &self._ctx.Q,
+                &_rnd.mbedtls_ctr_drbg_random, &__rng._ctx))
+            n = (_mpi.mbedtls_mpi_bitlen(&self._ctx.Q.X)+7)//8
```

Comment:
As mentioned in the comments to the PR, using an MPI from the beginning (```_public_key()```) should make the code simpler.